### DISCUSSION TITLE
Postgress 15 - A command is already in progress: SELECT problem

### DISCUSSION THREAD

**Discussion by goors** [linq2db/linq2db#4158] [General]
This is really stupid. I mean on my or Postgres end.

I have a simple query that goes like this:

`"select id from table where ids in(1, 1, 1)"`

translated to linq is `_db.table.Where(x => new[] {1, 1, 1}.Contains(x.id)).ToListAsync()`

**This error happens if I have two requests at the same time. Ui guys reported this.**

So I tried to replicate somehow this in .net core like this:

```
var ids = new List<long> { 1, 1, 1 };
var asyncTasks = new Task[3];
Parallel.ForEach(ids, new ParallelOptions { MaxDegreeOfParallelism = 5000 }, async (id, loopState, index) => {
     asyncTasks[index] = _db.table.Where(x => new[] {id}.Contains(x.Id)).ToListAsync()`
})
await Task.WhenAll(asyncTasks);
```

And the error is:

```
Npgsql.NpgsqlOperationInProgressException
A command is already in progress: SELECT
	x.Id
FROM
	table x
WHERE
	x.Id IN (1) --
```

Don't tell me this is impossible to resolve?

linq2db: 5.2.1
db: postgres 15

My first thought was, ok put it into transaction, like do not read unless pervious read is done (I could be wrong as hell for thinking this, please let me know). 

But using:

```
var tran = await _db.BeginTransactionAsync(IsolationLevel.ReadCommitted);  (btwl on Postgres default islocation leve is read commited)
// do query 
await tran.CommitAsync() 
```

add following error:

`The connection is already in state 'Executing'`

I could be wrong as hell here, but I can not believe that getting the same thing in parallel s not possible. It is like insane to think of, right?

**Comment by jods4** [comment]
Generally database connections are not multiplexed / reentrant / do not support multiple commands or queries running at the same time.
If you want to iterate over two result sets in parallel, you need two connections to your DB.

SQL Server has a proprietary feature called Multiple Active Result Sets (MARS) that enables this, but as far as I know Npgsql has no such feature, see https://github.com/npgsql/npgsql/issues/462.

**Reply by goors** [reply]
I am sorry but I think you missed the point:

1. user 1 comes to website
2. user 2 comes to website
3. they do getMeId(1) in matter in 100ms
4. one get the result and other get error select in progress?

I mean, this can not be right? @jods4

**Reply by jods4** [reply]
This is even worse. You're sharing objects across threads?!
No .net object is thread-safe/supports concurrent access unless specified and in particular nothing in ADO.NET is trade-safe.

Each request on your web server should open its own DB connection. They should share nothing and it will work.

**Reply by goors** [reply]
`_db` is registered as `services.TryAddTransient<GeoDb>()`; so I can use it everywhere and there is nothing special to it:

```
public class GeoDb
    {

        /// <summary>
        /// Initializes a new instance of this type.
        /// </summary>
        /// <param name="connectionString">Database connection string.</param>
        /// <remarks>
        /// </remarks>
        public GeoDb(IOptions<DataAccessOptions> options) : base(
            ProviderName.PostgreSQL95,
            options.Value?.ConnectionString
            )
        {}
}

```
And in other classes it is a part of constructor (where I need it).

What are you telling me, i should have in each service/class/repository open database/close database and do not add it as service. So it is a unknown number of services that are using database and in each i need to construct database, use it and close it?

**Reply by jods4** [reply]
A transient instance should be fine. 
I can't debug your code but the error message is clear:
> The connection is already in state 'Executing'

This indicates that you are trying to start an operation on a DbConnection that is already executing another request.
It typically happens in one of three ways:
- Async reentrancy: you make a new request before a previous async request on the same connection completes. Be sure to always await your async calls.
- Reader reentrancy: you start a new request while iterating over the DbReader of a previous one (iterating over selects in linq2db puts you in this situation), so the 2nd request starts when the 1st one is still in progress. Either use a second connection, or buffer the result set in a list before iteration.
- Concurrency: two threads are making concurrent calls on the same connection. You shouldn't share connections or use proper synchronization mechanisms (lock and co).

Happy debugging.

**Reply by goors** [reply]
Oh, that is part where I tried transaction. `The connection is already in state 'Executing'` I don't actually use transaction.  So the problem is the same.  @jods4

**Reply by jods4** [reply]
The original error is basically the same issue:
> A command is already in progress: 
> SELECT x.Id FROM table x WHERE x.Id IN (1) --

**Reply by goors** [reply]
Yes, that goes into the same problem. You can not register service once and use it. You need to do init class each time:

```
var cl = new SomeClass()
cl.method()

class SomeClass {
     private SomeDb db;
     SomeClass() {
         this.db = new Db(connection string)
      }
     method() { /// select id from table }
}
```

So as far as I can say, based on your comment that is the only way. Or in each method to have `var db = new Db() and then do db.where ....`

I mean if this is somethig Postgres requires this is really poor design. 1mil users requesting same id and some of them have result and some of them dont.... That is exactly the case here.

**Reply by goors** [reply]
I wanted to thank you for thoughts, but this is a case. Using beautiful language as a c# and power of dependency injections and when you have Postgres, to discover you can not really use DI is just sad ....

**Reply by viceroypenguin** [reply]
@goors: People use Postgres and Linq2db with Dependency Injection all the time. If you are encountering an error, there is likely an error in the configuration; you are receiving it from the DI incorrectly. 

The code you shared here (`this.db = new Db()`) indicates some unnecessary usage. The proper way to receive from the DI system is:
```cs
class SomeClass
{
  private readonly SomeDb _db;
  public SomeClass(SomeDb db)
  {
    _db = db;
  }
  public async Task<Entity> Method(int id)
  {
    await _db.Entities.SingleOrDefault(e => e.Id == id);
  }
}
```

Also, since `SomeClass` is part of the DI system, it must also be registered and received by consumers, rather than be a singleton. Usually as such:
```cs
public class SomeController
{
  private readonly SomeClass _service;
  public SomeController(SomeClass service)
  {
    _service = service;
  }

  [Get("/api/entity/{id}")]
  public async Task<Entity> GetEntity(int id)
  {
    return await _service.Method(id);
  }
}
```

There are other recommendations I would normally make here (return DTOs instead of Entity from the controller, etc.), but this should be enough for you to get started.

**Reply by goors** [reply]
I think you missed the point. 

1. user 1 comes to website
2. user 2 comes to website
3. they do getMeId(1) in matter in 100ms
4. one get the result and other get error select in progress?

And if you read my opening statement, I am using DI in the way that you wrote it, like in EXACTLY the same way. So really thank you for writing that code. 

In you `[Get("/api/entity/{id}")]` if that is fired in a matter of milliseconds you will have error`A command is already in progress: select` .

I tried to simulate that milliseconds in that `Task.WhenAll(tasks)` @viceroypenguin

**Reply by viceroypenguin** [reply]
@goors: 
The only way that two users separately coming to the website at the same time will encounter a `A command is already in progress: select` error is if they are sharing the same context. Obviously, I don't have access to your code, so I cannot tell you what is wrong with your code, but I have personally implemented using npgsql and mssql multiple times and do not get any errors with people accessing the site simultaneously.

Now. This code is wrong:
```cs
var ids = new List<long> { 1, 1, 1 };
var asyncTasks = new Task[3];
Parallel.ForEach(ids, new ParallelOptions { MaxDegreeOfParallelism = 5000 }, async (id, loopState, index) => {
     asyncTasks[index] = _db.table.Where(x => new[] {id}.Contains(x.Id)).ToListAsync()`
})
await Task.WhenAll(asyncTasks);
```

The reason it is wrong is because you are using the same `_db` across three parallel calls. You cannot use `_db` in multiple threads at the same time. This is true of all database libraries and ORMs. You will find this in mssql, you'll find it in npgsql, you'll find it in sqlite; you'll find it in Entity Framework, EFC, l2db, etc. A single database connection cannot be multiplexed.

In most cases, you will write the code as such:
```cs
foreach (var id in ids)
{
  var entity = await _db.Table.SingleOrDefault(x => x.Id == id);
  // do something with entity
}
```

Notice that the tasks never overlap, because we `await` anything going to the database. This means we're never trying to use the same connection for multiple queries _at the same time_.

If you really want to run in parallel, then you can do the following:
```cs
class SomeClass
{
  private readonly IServiceScopeFactory _scopeFactory
  public SomeClass(IServiceScopeFactory scopeFactory)
  {
    _scopeFactory = scopeFactory;
  }
  public async Task<Entity> Method(int id)
  {
    var ids = new List<long> { 1, 1, 1 };
    var entities = await Task.WhenAll(ids
      .Select(async id =>
      {
        using var scope = scopeFactory.CreateScope();
        var db = scope.GetRequiredService<SomeDb>();
        return await db.Table.SingleOrDefault(x => x.Id == id);
      });

    // do something with entities
  }
}
```

The reason this will work is because each task is getting it's own independent `SomeDb`, with it's own database connection. You're no longer trying to multiplex a single connection, you're making multiple parallel connections. 

Again, if you're ever getting a `A command is already in progress: select`, you're trying to use the same connection on multiple threads simultaneously.

**Reply by to11mtm** [reply]
The `Task.WhenAll` example is very clearly the 'wrong' thing, no matter what, as it is re-using a connection across tasks.

Without a more clear example of what your actual code is doing, the best suggestion I have is something like this, which is a pattern I've used for years to more or less minimize the risk of incorrect DI patterns or weird DI behaviors in certain libraries (e.x. Blazor:)

```
//Register as singleton
public class DataConnectionFactory
{
    private string _connectionString;
    //Ctor that takes connection string and/or DB type is left out here
    public DataConnection GetConnection()
    {
       return new DataConnection(ProviderName.PostgreSQL95, _connectionString);
    }
}

//Class that does data access:
public class SampleDataLayer
{
    private readonly DataConnectionFactory _factory;
    public SampleDataLayer(DataConnectionFactory factory)
    {
        _factory = factory;
    }
    public async Task<int> GetARecordCount()
    {
        using (var dc = _factory.GetConnection())
        {
            return await dc.GetTable<SomeTable>().CountAsync();
        }
    }
}
```

**Reply by goors** [reply]
So I got to the bottom of this @viceroypenguin. If `SomeClass` is registered as `Transient` and you have DI of `SomeDb` this will not work.

It will only work if your `SomeClass` is registered as Scoped.

I am thinking how expensive that is?

Is there a way to register `SomeDb` as Scoped. Would that be wiser? So every incoming HTTP request gets its own database connection, which is reused as much as possible for that single request.

I will try to make some minimal repository with this so everyone can test this.

**Reply by viceroypenguin** [reply]
> It will only work if your `SomeClass` is registered as Scoped.

This is correct. Both `SomeClass` and `SomeDb` should be registered as Scoped, so that they are instantiated for the connection.

**Reply by goors** [reply]
@viceroypenguin never thought of that. Silly of me, right. Thank you so much.
